### PR TITLE
feat: added endpoint to update a thread message

### PR DESCRIPTION
**Pull request by Shopiley**
Implemented update threads endpoint.

Note:
This is a temporary solution as it uses an O(n) algorithm while we review a better way to implement threads (possibly as a collection of it's own).
Also when testing the PUT /Update Message endpoint, the thread_id must be included in the request body.

@mikengr @zxenonx @AI-fae

### CODE REVIEW

File: backend/endpoints/threads.py [hunk 1]
```diff
@@ -5,9 +5,10 @@
 from schema.thread_response import ThreadResponse
 from fastapi.responses import JSONResponse
 from utils.centrifugo import Events, centrifugo_client
-from utils.message_utils import get_message, get_room_messages, update_message_threads as edit_message_threads
+from utils.message_utils import get_message, get_room_messages
 from utils.room_utils import get_org_rooms
 from utils.threads_utils import get_message_threads, add_message_to_thread_list
+from utils.threads_utils import  update_thread_message as edit_thread_message
```
Comment (FlorenceEgwu) [inline backend/endpoints/threads.py:11]: why is the alias neccessary?

File: backend/endpoints/threads.py [hunk 2]
```diff
@@ -168,42 +184,33 @@ async def update_threads_message(
     org_id: str,
     room_id: str,
     message_id: str,
+    thread_id: str,
```
Comment (FlorenceEgwu) [inline backend/endpoints/threads.py:187]: why not provide this as a path parameter for the purpose of consistence?

File: backend/requirements.txt [hunk 3]
```diff
@@ -14,7 +14,6 @@ colorama==0.4.4
 coverage==6.0.2
 distlib==0.3.3
 fastapi==0.70.0
-fastapi-pagination==0.9.1
```
Comment (FlorenceEgwu) [inline backend/requirements.txt:17]: how is this related to the current changes?

File: backend/requirements.txt [hunk 4]
```diff
@@ -46,7 +45,6 @@ python-dateutil==2.8.2
 python-dotenv==0.19.1
 python-multipart==0.0.5 
 PyYAML==6.0
-regex==2021.10.23
```
Comment (FlorenceEgwu) [inline backend/requirements.txt:49]: why???

File: backend/utils/threads_utils.py [hunk 5]
```diff
@@ -56,8 +56,40 @@ async def add_message_to_thread_list(org_id, room_id, message_id, request: Threa
...
+         payload (dict[str, Any]): The new data.
+         message (dict[str, Any]): The message to be edited that has been loaded from the database
+     Returns:
+         dict[str, Any]: The response returned by DataStorage's update method.
+     """
+    threads = message["threads"]  # a list of thread objects
+
+    # to edit a thread object in the list of threads
+    for index in range(len(threads)):
+
```
Comment (FlorenceEgwu) [inline backend/utils/threads_utils.py:83]: Like you already mentioned in your PR description, this is a potential problem. we can't be looping through 100 threads just to update 1

File: backend/utils/threads_utils.py [hunk 6]
```diff
@@ -56,8 +56,40 @@ async def add_message_to_thread_list(org_id, room_id, message_id, request: Threa
...
+            threads[index]["edited"] = True
+
+            del message["_id"]
+
+            db = DataStorage(org_id)
+            return await db.update(settings.MESSAGE_COLLECTION, message_id, message)
+
+    raise HTTPException(
+        status_code=status.HTTP_404_NOT_FOUND, detail="Thread message not found or sender is not authorised to edit this message"
+    )
```
Comment (FlorenceEgwu) [inline backend/utils/threads_utils.py:95]: I think it will be best if 2 different exceptions are raised here. One it a 404 exception while the other is a 403 (where the sender is not the same) exception.
Best to keep things explicit for readability in the near future.

**Review by FlorenceEgwu — CHANGES_REQUESTED** [review]

**Comment by FlorenceEgwu** [comment]
Good Job Sope!
I have left a few comments.

BTW, is there anything particularly stopping us from restructuring things and implementing this threads feature correctly right away?

File: backend/endpoints/threads.py (hunk 1 above)
Comment (zxenonx) [reply backend/endpoints/threads.py:11]: @Shopiley, please rename the endpoint function to **update_thread_message** and that would make sense.

Comment (zxenonx) [inline backend/endpoints/threads.py:11]: Kindly merge these imports.

File: backend/endpoints/threads.py (hunk 2 above)
Comment (zxenonx) [reply backend/endpoints/threads.py:187]: Yes please, the id of the resource to be edited must be provided in the route. @Shopiley

Comment (zxenonx) [reply backend/endpoints/threads.py:187]: Also please add a docstring to the function.

File: backend/endpoints/threads.py [hunk 7]
```diff
@@ -168,42 +184,33 @@ async def update_threads_message(
...
-    if not added_thread_message:
-        raise HTTPException(
-            status_code=status.HTTP_424_FAILED_DEPENDENCY,
-            detail={"thread not added": added_thread_message},
-        )
+    payload = request.dict(exclude_unset=True)
 
-    # payload["edited"] = True
-    loadedMessage.update(thread_to_add)
+    edited_thread_message = await edit_thread_message(org_id, message_id, thread_id, payload, message)
```
Comment (zxenonx) [inline backend/endpoints/threads.py:201]: This variable is unused...
You should check whether the thread msg has been successfully edited or not.

File: backend/endpoints/threads.py [hunk 8]
```diff
@@ -168,42 +184,33 @@ async def update_threads_message(
...
         room_id,
         Events.MESSAGE_UPDATE,
-        loadedMessage
+        message
     )
 
     return JSONResponse(
         content=ResponseModel.success(
-            data=loadedMessage, message="Thread Added"),
+            data=message, message="Thread Edited"),
```
Comment (zxenonx) [inline backend/endpoints/threads.py:213]: Thanks for the correction here.
loadedMessage is not a valid variable in Python.

File: backend/utils/message_utils.py [hunk 9]
```diff
@@ -179,18 +179,19 @@ async def update_message(
...
-        message_id (str): The id of the message to be edited.
-        message (dict[str, Any]): The new data.
-    Returns:
-        dict[str, Any]: The response returned by DataStorage's update method.
-    """
-
-    db = DataStorage(org_id)
-
-    return await db.update(settings.MESSAGE_COLLECTION, message_id, message)
+# async def update_thread_message(
```
Comment (zxenonx) [inline backend/utils/message_utils.py:182]: If it will no longer be used, we can just delete it.

File: backend/utils/threads_utils.py [hunk 10]
```diff
@@ -56,8 +56,40 @@ async def add_message_to_thread_list(org_id, room_id, message_id, request: Threa
...
-   
+
     response = await update_message(org_id, message_id, message)
-    
+
     return response, thread_message
 
+
+async def update_thread_message(
+    org_id: str, message_id: str, thread_id: str, payload: dict[str, Any], message: dict[str, Any]
```
Comment (zxenonx) [inline backend/utils/threads_utils.py:66]: Do we still need **message** when **message_id** is already provided?
5 params, This is starting to be a lot.

File: backend/utils/threads_utils.py (hunk 5 above)
Comment (zxenonx) [reply backend/utils/threads_utils.py:83]: We need to find a better implementation here.
But just a quick suggestion, in this situation it's better to use **enumerate** instead of **range+len**.

https://themeptation.medium.com/why-use-enumerate-instead-of-range-in-your-pythons-loops-d33bfd9c7531

**Review by zxenonx — CHANGES_REQUESTED** [review]
Good job @Shopiley.
Kindly go through the comments I left.
Thanks.

**Comment by Shopiley** [comment]
Thanks for comments! I've made [another PR](https://github.com/zurichat/zc_messaging/pull/213) implementing an improved solution. I also made corrections based on your comments here.

File: backend/utils/threads_utils.py (hunk 6 above)
Comment (Shopiley) [reply backend/utils/threads_utils.py:95]: Done
